module/VuFindTheme/src/VuFindTheme/ResourceContainer.php

```diff
@@ -130,13 +130,100 @@ public function addCss($css)
     public function addJs($js)
     {
         if (!is_array($js) && !is_a($js, 'Traversable')) {
-            $js = [$js];
+            $this->addJsEntry($js);
+        } elseif (isset($js['file'])) {
+            $this->addJsEntry($js);
+        } elseif (isset($js[0])) {
+            foreach ($js as $current) {
+                $this->addJsEntry($current);
+            }
+        } else {
+            trigger_error("Invalid JS entry format: " . print_r($js, true));
+        }
+    }
+
+    /**
+     * Helper function for adding a Javascript file.
+     *
+     * @param string|array $jsEntry Entry to add, either as string with path
+     * or array with additional properties.
+     *
+     * @return void
+     */
+    protected function addJsEntry($jsEntry)
+    {
+        if (!is_array($jsEntry)) {
+            $parts = $this->parseSetting($jsEntry);
+            if (count($parts) == 1) {
+                $jsEntry = ['file' => $jsEntry];
+            } else {
+                $jsEntry = [
+                    'file' => $parts[0],
+                    'attributes' => ['conditional' => trim($parts[1])],
+                ];
+            }
         }
-        foreach ($js as $current) {
-            $this->js[] = $current;
+        if (!isset($jsEntry['position'])) {
+            $jsEntry['position'] = 'header';
+        }
+
+        if (!in_array($jsEntry, $this->js, true)) {
+            $this->insertEntry($jsEntry, $this->js);
         }
     }
 
+    /**
+     * Helper function to insert an entry to an array,
+     * also considering prioroty and dependancy, if existing.
+     *
+     * @param array $entry The entry to insert.
+     * @param array $array The array into which the entry shall be inserted.
+     *
+     * @return void
+     */
+    protected function insertEntry($entry, &$array)
+    {
+        if (isset($entry['priority']) || isset($entry['dependency'])) {
+            for ($i=0; $i < count($array);++$i) {
+                if (isset($entry['priority'])) {
+                    $currentPriority = $array[$i]['priority'] ?? null;
+                    if (!isset($currentPriority)
+                        || $currentPriority > $entry['priority']
+                    ) {
+                        $this->insertEntryAtPosition($entry, $i, $array);
+                        return;
+                    }
+                } elseif (isset($entry['dependency'])) {
```

Comment:
If we want to keep this simple, I wonder if `load_after` would be a better name for the setting than `dependency` -- `dependency` implies to me more complex/sophisticated dependency resolution, and I don't want us to end up having to reinvent Composer inside VuFind's theme system. :-)

A couple of other questions/concerns:

1.) Do we need to worry about dependencies on multiple files? If one file needs to be loaded after two other files, and we can't predict the order of those files being loaded, how do we deal with that situation?

2.) Do we need a "load_before" setting as a counterpart to "load_after"?

3.) What do we do if the file we're supposed to depend upon is not already present in the Javascript array? Do we throw an exception? If we're not guaranteeing the order if the input, I don't think we can assume that the dependency is already available at the time we load the dependent file.

(And as a side note, all of these types of questions are the reason I tend to prefer a priority system, because it's simple and unambiguous... but of course the downside is that it requires management of priority levels... I just don't think that's likely to be too complex for VuFind, if we just use something like priority 0 for core libraries like jQuery and Bootstrap, priority 50 for VuFind-specific support code, and priority 100 for anything more specific than that... and then local custom code could use higher numbers as needed. I'm not opposed to continuing to work through the issues of a before/after/dependency system, but just warning that I think it will take significant work to cover all the possible edge cases.)

Reply:
OK, i just renamed "dependency" to "load_after". Regarding your other questions:

1) We should always be able to predict the order of the other files by adding clear priorities or dependencies.

2) I think that's not really necessary. If there might be a case, this could be indeed solved by using a priority instead.

3) I guess the order should be guaranteed, else it would be necessary to implement a 2-pass-sorting algorithm. So i just added an exception if the "load_after" dependency could not be found.

Reply:
Thanks! This sounds good enough for now, and we can add more complexity if the need arises. :-)